contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,633 @@
+
+### Recommender ###
+Recommender is the main component of the VPA. It is responsible for
+computing recommended resources. On startup the recommender fetches
+historical resource utilization of all Pods (regardless of whether
+they use VPA) together with the history of Pod OOM events from the
+History Storage. It aggregates this data and keeps it in memory.
+
+During normal operation the recommender consumes real time updates of
+resource utilization and new events via the Metrics API from
+the [Metrics Server](https://github.com/kubernetes-incubator/metrics-server).
+Additionally it watches all Pods and all VPA objects in the
+cluster. For every Pod that is matched by some VPA selector the
+Recommender computes the recommended resources and sets the
+recommendation on the VPA object.
+
+It is important to realize that one VPA object has one recommendation.
+The user is expected to use one VPA to control Pods with similar
+resource usage patterns, typically a group of replicas or shards of
+a single workload.
+
+The Recommender acts as an
+[extension-apiserver](https://kubernetes.io/docs/concepts/api-extension/apiserver-aggregation/),
+exposing a synchronous RPC interface that takes a Pod Spec and the Pod metadata
+and returns recommended resources.
+
+#### Recommender API ####
+
+```GET /recommendedPodResources```
+
+Request body:
+```go
+type GetRecommendedPodResourcesRequest {
+  PodMetadata metav1.ObjectMeta
+  PodSpec core.PodSpec
+}
+```
+The response:
+```go
+type GetRecommendedPodResourcesReponse {
+  Recommendation autoscaler.RecommendedPodResources
+}
+```
+
+Notice that this API method may be called for an existing Pod, as well as for a
+yet-to-be-created Pod.
+
+### Updater ###
+VPA Updater is a component responsible for applying recommended resources to
+existing Pods. In MVP (until in-place updates of Pod resources are available)
+this means evicting Pods whose recommended resources significantly diverge from
+actual resources, in order to have them recreated with the recommended
+resources.
+
+While terminating Pods is disruptive and generally undesired, it is sometimes
+justified in order to (1) avoid CPU starvation (2) reduce the risk of correlated
+OOMs across multiple Pods at random time or (3) save resources over long periods
+of time.
+
+Apart from its own policy on how often a Pod can be evicted, the Updater also
+respects the Pod disruption budget, by using Eviction API to evict pods.
+
+The Updater only touches Pods that (1) point to a VPA with updatePolicy.mode set
+to `"Auto"` and (2) belong to a Replica Set.
+
+The Updater will also need to understand how to adjust the recommendation before
+applying it to a Pod, based on the current state of the cluster (e.g. quota,
+space available on nodes or other scheduling constraints).
+Otherwise it may deschedule a Pod permanently. This mechanism is not yet
```

Comment:
Does the Updater use the cached recommendation?  Does it update the recommendation itself?  I think you need to clarify the role of the cached recommendation and how the updater interacts with the reccomender.

Reply:
Done.
Updater calls the Recommender API.
The cached recommendations are only used by the admission controller in case the Recommender is down, so that it's not a critical dependency for creating new pods.